solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudHttp2SolrClient.java

```diff
@@ -56,10 +56,14 @@ protected CloudHttp2SolrClient(Builder builder) {
     super(builder.shardLeadersOnly, builder.parallelUpdates, builder.directUpdatesToLeadersOnly);
     this.clientIsInternal = builder.httpClient == null;
     this.myClient = createOrGetHttpClientFromBuilder(builder);
-    this.stateProvider =
-        builder.zkHosts.isEmpty()
-            ? createHttp2ClusterStateProvider(builder.solrUrls, myClient)
-            : createZkClusterStateProvider(builder);
+
+    try {
+      ClusterStateProviderFactory factory = new ClusterStateProviderFactory(builder, myClient);
+      this.stateProvider = factory.create();
```

Comment:
Looking at this; I have to wonder if we really need a Factory class at all or do we just need a factory *method*.  Maybe on ClusterStateProvider or the builder.  Factory classes are useful if it's pluggable or configurable but that's not the case here.

Reply:
There were already two Factory method there -- one for ZkClientClusterStateProvider and another HttpCSP. My idea is to encapsulate that and the new custom logic into one unit. So that it should be easier for others to read this code. Yes I can move this class to builder so that we deal with stateProvider in the builder only rather than moving it to the main class.